### PR TITLE
[#45] scheduler to refresh registered repository resources

### PR DESCRIPTION
**Pull request by pkdash**

### CODE REVIEW

File: triggers/scheduler.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    db = AsyncIOMotorClient(settings.db_connection_string)[settings.database_name]
+    await init_beanie(database=db, document_models=[Submission, DatasetMetadataDOC])
+
+    async for submission in Submission.find(Submission.repository != None):
+        try:
+            dataset = await DatasetMetadataDOC.get(submission.identifier)
+            if dataset is None:
+                # TODO: delete submission record?
+                # await submission.delete()
+                # logger.warning(f"Deleting Submission {submission.url} as it is not in catalog")
```
Comment (sblack-usu) [inline triggers/scheduler.py]: This is a case that should not happen and indicates that there is an issue with the data integrity.  I do not think we should delete the submission.  I do want to be aware if this case happens.

I also don't know if we want to delete the discovery record because if we do get to this point and the discovery record exists, we can repair the submission by copying the discovery record to the catalog.

Long story short, I think we should log a warning then move on without making any changes to the collections.

File: api/models/user.py [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
 from beanie import Document, Link, PydanticObjectId
 from pydantic import HttpUrl
 
-from api.adapters.utils import RepositoryType
+if TYPE_CHECKING:
+    # this avoids circular imports
+    from api.adapters.utils import RepositoryType
 
 
 class Submission(Document):
```
Comment (sblack-usu) [inline api/models/user.py:12]: I think you asked about this previously, the submissions and catalog entries are confusing.  To simplify, I think we can drop the submissions and use the catalog entry in its place.  When we need the submissions, we can just create the json on the fly to respond with in the api.

What do you think? @pkdash

Comment (sblack-usu) [reply api/models/user.py:12]: I'm looking at this more and the submitted date would have to be added to the catalog entry.  I think this is too large of a change to really entertain at the moment. I'm going to have to think this all through for a longer time.

File: triggers/scheduler.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            if updated_dataset is not None:
+                # update catalog record
+                await dataset.set(updated_dataset.dict(exclude_unset=True, by_alias=True))
+
+                # update submission record
+                dataset = await DatasetMetadataDOC.get(submission.identifier)
+                updated_submission = dataset.as_submission()
+                updated_submission.repository_identifier = submission.repository_identifier
+                updated_submission.repository = submission.repository
+                await submission.set(updated_submission.dict(exclude_unset=True))
```
Comment (sblack-usu) [inline triggers/scheduler.py]: Does this preserve the original submission date?  I think exclude_unset would exclude a property with a datetime factory, but I just want to confirm.

**Review by sblack-usu — CHANGES_REQUESTED** [review]
This is all fine, the only change I want to ask for is updating the scheduler job to log a warning and not delete the discovery record in the case that a submission does not have a catalog entry.

File: api/models/user.py (hunk 2 above)
Comment (pkdash) [reply api/models/user.py:12]: > I think you asked about this previously, the submissions and catalog entries are confusing. To simplify, I think we can drop the submissions and use the catalog entry in its place. When we need the submissions, we can just create the json on the fly to respond with in the api.

If we can drop submission collection that would be better. Can we store additional attributes in catalog record which may not be schema.org compatible like submission date, and repository record registration related attributes (repository name and repository record id etc.)? Submission allows us to store any arbitrary attributes we may need. 

> I'm looking at this more and the submitted date would have to be added to the catalog entry. I think this is too large of a change to really entertain at the moment. I'm going to have to think this all through for a longer time.

Agree we need bit longer time to think through this scheduler pipeline.

File: triggers/scheduler.py (hunk 3 above)
Comment (pkdash) [reply triggers/scheduler.py]: Good question/catch. It does not preserver the original submission date. Need to do this:
`updated_submission.submitted = submission.submitted`

**Comment by sblack-usu** [comment]
Does this need to be rebased?  It looks like it contains some work that is already in develop?

**Comment by pkdash** [comment]
I did merge the latest develop.

**Comment by pkdash** [comment]
@sblack-usu Can you review again? Thanks.

**Review by sblack-usu — APPROVED** [review]
